src/producer/vector_producer.rs

```diff
@@ -1,57 +1,108 @@
 }
 
 impl VectorResultsProducer {
     pub fn new(
         topic_name: &str,
         endpoint: String,
         vector_batch_size: usize,
+        retry_vector_errors_forever: bool,
+        region: String,
     ) -> (Self, JoinHandle<()>) {
         let schema = sentry_kafka_schemas::get_schema(topic_name, None).unwrap();
         let client = Client::new();
+        let pending_items = Arc::new(AtomicUsize::new(0));
+
+        let config = VectorRequestWorkerConfig {
+            vector_batch_size,
+            endpoint,
+            retry_vector_errors_forever,
+            region,
+        };
 
         let (sender, receiver) = mpsc::unbounded_channel();
-        let worker = Self::spawn_worker(vector_batch_size, client, endpoint, receiver);
+        let worker = Self::spawn_worker(config, client, receiver, pending_items.clone());
 
-        (Self { schema, sender }, worker)
+        (
+            Self {
+                schema,
+                sender,
+                pending_items,
+            },
+            worker,
+        )
     }
 
     fn spawn_worker(
-        vector_batch_size: usize,
+        config: VectorRequestWorkerConfig,
         client: Client,
-        endpoint: String,
         mut receiver: UnboundedReceiver<Vec<u8>>,
+        pending_items: Arc<AtomicUsize>,
     ) -> JoinHandle<()> {
         tracing::info!("vector_worker.starting");
 
         tokio::spawn(async move {
-            let mut batch = Vec::with_capacity(vector_batch_size);
+            let mut batch = Vec::with_capacity(config.vector_batch_size);
 
             while let Some(json) = receiver.recv().await {
+                pending_items.fetch_sub(1, Ordering::SeqCst);
+                metrics::gauge!("producer.pending_items")
+                    .set(pending_items.load(Ordering::SeqCst) as f64);
+
                 batch.push(json);
 
-                if batch.len() < vector_batch_size {
+                if batch.len() < config.vector_batch_size {
                     continue;
                 }
 
                 let batch_to_send =
-                    std::mem::replace(&mut batch, Vec::with_capacity(vector_batch_size));
-                if let Err(e) = send_batch(batch_to_send, client.clone(), endpoint.clone()).await {
+                    std::mem::replace(&mut batch, Vec::with_capacity(config.vector_batch_size));
+                if let Err(e) = {
+                    let start = std::time::Instant::now();
+                    let result = send_batch(
+                        batch_to_send,
+                        client.clone(),
+                        config.endpoint.clone(),
+                        config.retry_vector_errors_forever,
+                    )
+                    .await;
+                    metrics::histogram!("vector_producer.send_batch.duration", "uptime_region" => config.region.clone(), "histogram" => "timer").record(start.elapsed().as_secs_f64());
+                    result
+                } {
```

Comment:
nit: Having this whoel block inside the if is getting a bit unweildy, I think we could just do the `if` on the result itself

```rs
if let Err(e) = result {
   // ...
}
```